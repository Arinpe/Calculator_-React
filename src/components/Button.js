import PropTypes from 'prop-types';

export default function Button({
  name, onClick, color, wide,
}) {
  return (
    <button
      className={`btn ${color ? 'operatorBtn' : ''} ${wide ? 'wide' : ''}`}
      type="button"
      value={name}
      onClick={onClick}
    >
      {name}
    </button>
  );
}

Button.defaultProps = {
  color: null,
  wide: null,
};

Button.propTypes = {
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  color: PropTypes.bool,
  wide: PropTypes.bool,
};
