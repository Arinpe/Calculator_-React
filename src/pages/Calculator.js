import React from 'react';
import Display from '../components/Display';
import ButtonPanel from '../components/ButtonPanel';
import calculate from '../logic/calculate';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      total: null,
      next: null,
    };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(ButtonName) {
    this.setState(data => calculate(data, ButtonName));
  }

  render() {
    const { total, next } = this.state;
    const result = next || total || '0';

    return (
      <div className="seperate">

        <h1> LET S DO SOME MATH</h1>
        <div>
          <Display total={result.toString()} />
          <ButtonPanel clickHandler={this.handleClick} />
        </div>
      </div>
    );
  }
}

export default Calculator;
