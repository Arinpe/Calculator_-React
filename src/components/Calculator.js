import React, { Component } from 'react';
import ButtonPannel from './ButtonPanel';
import Display from './Display';
import Button from './Button';
import calculate from '../logic/calculate';

const buttons = [
  ['AC', '+/-', '%', '/'],
  ['9', '8', '7', '*'],
  ['6', '5', '4', '+'],
  ['3', '2', '1', '-'],
  ['0', '.', '='],
];
class Calculator extends Component {
  constructor() {
    super();
    this.state = {
      total: null,
      prev: '0',
      next: '0',
      operator: null,
    };
  }

  handledisplay = (prev, next, tot, op) => {
    let ans;
    if (tot) {
      ans = tot;
    } else if (!op) {
      ans = prev;
    } else {
      ans = next;
    }
    return ans;
  };

  handleClick = e => {
    const {
      total, next, operator, prev,
    } = this.state;
    const res = calculate({
      total, next, operator, prev,
    }, e.target.value);

    this.setState({
      ...res,
    });
  };

  render() {
    const {
      prev, next, total, operator,
    } = this.state;
    return (
      <>
        <div className="App">
          <div className="App__calculator">
            <Display current={this.handledisplay(prev, next, total, operator)} />
            <ButtonPannel>
              {buttons.map(group => (
                <div key={group} className="row">
                  {group.map(button => {
                    const operators = ['+', '-', '*', '/', '='];
                    const isOperator = operators.includes(button);
                    return (
                      <Button
                        onClick={this.handleClick}
                        key={button}
                        name={button}
                        color={isOperator}
                        wide={button === '0'}
                      />
                    );
                  })}
                </div>
              ))}
            </ButtonPannel>
          </div>
        </div>
      </>
    );
  }
}

export default Calculator;
