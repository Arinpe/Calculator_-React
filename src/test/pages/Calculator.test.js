import React from 'react';
import render from 'react-test-renderer';
import Calculator from '../../components/Calculator';
import Button from '../../components/Button';

describe('Calculator', () => {
  const page = render.create(<Calculator />);
  const pageInstance = page.root;

  test('should render without crashing', () => {
    expect(page).toMatchSnapshot();
  });

  test('should display the div with App class', () => {
    expect(pageInstance.findAllByProps({ className: 'App' }).length).toEqual(1);
  });

  test('should render the calculator container', () => {
    expect(pageInstance.findAllByProps({ className: 'App__calculator' }).length).toEqual(1);
  });

  test('should render the buttons', () => {
    expect(pageInstance.findAllByType(Button).length).toEqual(19);
  });
});
