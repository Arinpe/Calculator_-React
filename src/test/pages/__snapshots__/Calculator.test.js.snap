// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should render without crashing 1`] = `
<div
  className="App"
>
  <div
    className="App__calculator"
  >
    <div
      className="display"
    >
      <div
        className="displayupper"
      >
        
      </div>
      <div
        className="displaylower"
      >
        0
      </div>
    </div>
    <div
      className="button_panel"
    >
      <div
        className="row"
      >
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="AC"
        >
          AC
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="+/-"
        >
          +/-
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="%"
        >
          %
        </button>
        <button
          className="btn operatorBtn "
          onClick={[Function]}
          type="button"
          value="/"
        >
          /
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="9"
        >
          9
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="8"
        >
          8
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="7"
        >
          7
        </button>
        <button
          className="btn operatorBtn "
          onClick={[Function]}
          type="button"
          value="*"
        >
          *
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="6"
        >
          6
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="5"
        >
          5
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="4"
        >
          4
        </button>
        <button
          className="btn operatorBtn "
          onClick={[Function]}
          type="button"
          value="+"
        >
          +
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="3"
        >
          3
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="2"
        >
          2
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="1"
        >
          1
        </button>
        <button
          className="btn operatorBtn "
          onClick={[Function]}
          type="button"
          value="-"
        >
          -
        </button>
      </div>
      <div
        className="row"
      >
        <button
          className="btn  wide"
          onClick={[Function]}
          type="button"
          value="0"
        >
          0
        </button>
        <button
          className="btn  "
          onClick={[Function]}
          type="button"
          value="."
        >
          .
        </button>
        <button
          className="btn operatorBtn "
          onClick={[Function]}
          type="button"
          value="="
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
